### PR TITLE
consider region provder in client builders when possible

### PR DESCRIPTION
**Pull request by n0tl3ss**
Adds RegionProvider to the clients if possible to determine the region.

### CODE REVIEW

**Review by graemerocher — APPROVED** [review]

File: oraclecloud-common/src/main/java/io/micronaut/oraclecloud/core/sdk/AbstractSdkClientFactory.java [hunk 1]
```diff
@@ -71,12 +72,14 @@ protected AbstractSdkClientFactory(
      *
      * @param clientBuilder The builder for client
      * @param authenticationDetailsProvider The authentication details provider
+     * @param regionProvider The region provider
      * @return The client to build
      */
     protected abstract @NonNull T build(
         @NonNull B clientBuilder,
-        @NonNull AbstractAuthenticationDetailsProvider authenticationDetailsProvider
```
Comment (sdelamo) [inline oraclecloud-common/src/main/java/io/micronaut/oraclecloud/core/sdk/AbstractSdkClientFactory.java:78]: we should keep and deprecate the old method

File: oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/factory/ObjectStorageFactorySpec.groovy [hunk 2]
```diff
@@ -67,4 +72,16 @@ class ObjectStorageFactorySpec extends Specification {
         }
     }
 
+    @Singleton
+    @BootstrapContextCompatible
+    @Primary
```
Comment (sdelamo) [inline oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/factory/ObjectStorageFactorySpec.groovy]: I think `@Primary` is unnecessary. It suffices with the `@Replaces`.

File: oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/sdk/VaultRegionProviderMock.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import com.oracle.bmc.Region;
+import com.oracle.bmc.auth.RegionProvider;
+import io.micronaut.context.annotation.BootstrapContextCompatible;
+import io.micronaut.context.annotation.Primary;
+import io.micronaut.context.annotation.Property;
+import io.micronaut.context.annotation.Replaces;
+import io.micronaut.context.annotation.Requires;
+import jakarta.inject.Singleton;
+
+@Primary
```
Comment (sdelamo) [inline oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/sdk/VaultRegionProviderMock.java]: You don't need the `@Primary` annotation. `@Replaces` suffices.  

```suggestion
```

**Review by sdelamo — APPROVED** [review]

File: oraclecloud-common/src/main/java/io/micronaut/oraclecloud/core/sdk/AbstractSdkClientFactory.java (hunk 1 above)
Comment (sdelamo) [reply oraclecloud-common/src/main/java/io/micronaut/oraclecloud/core/sdk/AbstractSdkClientFactory.java:78]: I have annotated the factory as @Internal instead. I think it is an error that it was not annotated as `@Internal`

File: oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/factory/ObjectStorageFactorySpec.groovy (hunk 2 above)
Comment (sdelamo) [reply oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/factory/ObjectStorageFactorySpec.groovy]: I have removed the `@Primary` annotation here: https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/1034/commits/b98c8f521a68fb1ad2645181e63ee10be63d47c6

File: oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/sdk/VaultRegionProviderMock.java (hunk 3 above)
Comment (sdelamo) [reply oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/sdk/VaultRegionProviderMock.java]: I have removed the `@Primary` annotation here: https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/1034/commits/b98c8f521a68fb1ad2645181e63ee10be63d47c6

**Pull request closed without merging**